### PR TITLE
Move TCL test integration/replication to Go case

### LINKED ISSUE #948: Move TCL test integration/replication to Go case

### LINKED ISSUE #996: Fix didn't send replication master auth if it's created in flight
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-kvrocks/issues) and found no similar issues.

### Version

21b92a9f11cdd479498e592e500c29d137d8eec1

### Minimal reproduce step

1. Start `serverA` and `serverB` without `requirepass`
2. `ServerB` slave of `slaveA`
3. Use the config set command to create requirepass/master for the `serverA` and `serverB`
4. Client kill type slave in the master(`serverA`)

### What did you expect to see?

The slave(`serverB`) is disconnected from the master `slaveA`

### What did you see instead?

The slave(`serverB`) is re-connected with the master `slaveA` normally

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by git-hulk**
* This closes #948.
* This closes #996.

This PR also fixes the issue that didn't send the replication master auth if it's created in flight, for more information can see: #996 
    
The root cause is we determined whether to send auth command or not on creating the replication state machine. If we add the master auth after creating the replication thread, it will never send auth command without recreating the replication state machine.

### CODE REVIEW

**Comment by git-hulk** [comment]
@ShooterIT @caipengbo Can have a look the bug which mentioned in #996

File: tests/gocase/unit/introspection/replication_test.go [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+	slaveClient := slave.NewClient()
+	defer func() { require.NoError(t, slaveClient.Close()) }()
+
+	util.SlaveOf(t, slaveClient, master)
+	util.WaitForSync(t, slaveClient)
+
+	t.Run("Master doesn't pause replicating with replicas, #346", func(t *testing.T) {
+		ctx := context.TODO()
+		// In #346, we find a bug, if one command contains more than special
+		// number updates, master won't send replication stream to replicas.
```
Comment (tisonkun) [inline tests/gocase/unit/introspection/replication_test.go]: If it refers to Redis's issue 346, we may specify it clearly.

**Comment by caipengbo** [comment]
@git-hulk I have no problem with the #996 bug fix. However, there are two things in this PR, I think the title should be modified or split into two PR.

File: tests/gocase/unit/introspection/replication_test.go [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+		masterClient.HSet(ctx, "myhash", 0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9,
+			"a", "a", "b", "b", "c", "c", "d", "d", "e", "e", "f", "f", "g", "g", "h", "h", "i", "i", "j", "j", "k", "k")
```
Comment (tisonkun) [inline tests/gocase/unit/introspection/replication_test.go]: Perhaps using the variant:

```go
HSet("myhash", map[string]interface{}{"key1": "value1", "key2": "value2"})
```

File: tests/gocase/unit/introspection/replication_test.go [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+	slave := util.StartServer(t, map[string]string{})
+	defer slave.Close()
+	slaveClient := slave.NewClient()
+	defer func() { require.NoError(t, slaveClient.Close()) }()
+
+	util.SlaveOf(t, slaveClient, master)
+	util.WaitForSync(t, slaveClient)
+
+	t.Run("Slave can re-sync with master after password change", func(t *testing.T) {
+		ctx := context.TODO()
```
Comment (tisonkun) [inline tests/gocase/unit/introspection/replication_test.go]:
```suggestion
		ctx := context.Background()
```

nit: consistently using `context.Background()`.

File: tests/gocase/unit/introspection/replication_test.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+
+		killedSlaveCount, err := masterClient.ClientKillByFilter(ctx, "type", "slave").Result()
+		require.NoError(t, err)
+		require.Greater(t, killedSlaveCount, int64(0))
+
+		util.WaitForSync(t, slaveClient)
+		masterReplicationInfo = masterClient.Info(ctx, "replication").String()
+		require.Contains(t, masterReplicationInfo, "role:master")
+		require.Contains(t, masterReplicationInfo, slave.Host())
+		require.Contains(t, masterReplicationInfo, strconv.Itoa(int(slave.Port())))
```
Comment (tisonkun) [inline tests/gocase/unit/introspection/replication_test.go]: This seems not quite equal to:

```tcl
            set ret [$master client list]
            set lines [split $ret "\n"]
            global conn_addr
            foreach  line $lines {
                if {[string match "*cmd=psync*" $line]} {
                    set list [split $line " "]
                    foreach str $list {
                        if {[string match "*addr=*" $str]} {
                            set conn_addr [string range $str 5 end]
                        }
                    }
                }
            }

            $master client kill $conn_addr
            wait_for_sync $slave
            catch {$master info replication} var
            assert_match {*role:master*[$slave_ip]*[$slave_port]*} $var
```

any reason?

**Review by tisonkun — CHANGES_REQUESTED** [review]
Great work! Comments inline.

**Comment by git-hulk** [comment]
> @git-hulk I have no problem with the #996 bug fix. However, there are two things in this PR, I think the title should be modified or split into two PR.

it's a corner case bug and also related to testing cases in this PR, so I think it's fine to fix in the same PR.

File: tests/gocase/unit/introspection/replication_test.go (hunk 1 above)
Comment (git-hulk) [reply tests/gocase/unit/introspection/replication_test.go]: Sure, it's just copied from previous test cases, we should claim it clearly.

File: tests/gocase/unit/introspection/replication_test.go (hunk 4 above)
Comment (git-hulk) [reply tests/gocase/unit/introspection/replication_test.go]: Kvrocks supported using the type to kill the slave connection, the old implementation is a bit wordy and tricky.

Comment (tisonkun) [reply tests/gocase/unit/introspection/replication_test.go]: Get it.

File: tests/gocase/unit/introspection/replication_test.go [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package introspection
```
Comment (tisonkun) [inline tests/gocase/unit/introspection/replication_test.go]: I suggest this case should be under `gocase/integration/replication/replication_test.go`.

**Review by tisonkun — COMMENTED** [review]
Unstable case. Please take a look.

```
--- FAIL: TestReplicationChangePassword (2.74s)
    --- FAIL: TestReplicationChangePassword/Slave_can_re-sync_with_master_after_password_change (0.10s)
        replication_test.go:346: 
            	Error Trace:	/Users/runner/work/incubator-kvrocks/incubator-kvrocks/tests/gocase/unit/introspection/replication_test.go:346
            	Error:      	"info replication: # Replication\r\nrole:master\r\nconnected_slaves:1\r\nmaster_repl_offset:0\r\n" does not contain "127.0.0.1"
            	Test:       	TestReplicationChangePassword/Slave_can_re-sync_with_master_after_password_change
```

File: tests/gocase/unit/introspection/replication_test.go (hunk 5 above)
Comment (git-hulk) [reply tests/gocase/unit/introspection/replication_test.go]: okay, will move it.

File: tests/gocase/unit/introspection/replication_test.go (hunk 3 above)
Comment (git-hulk) [reply tests/gocase/unit/introspection/replication_test.go]: > Unstable case. Please take a look.
> 
> ```
> --- FAIL: TestReplicationChangePassword (2.74s)
>     --- FAIL: TestReplicationChangePassword/Slave_can_re-sync_with_master_after_password_change (0.10s)
>         replication_test.go:346: 
>             	Error Trace:	/Users/runner/work/incubator-kvrocks/incubator-kvrocks/tests/gocase/unit/introspection/replication_test.go:346
>             	Error:      	"info replication: # Replication\r\nrole:master\r\nconnected_slaves:1\r\nmaster_repl_offset:0\r\n" does not contain "127.0.0.1"
>             	Test:       	TestReplicationChangePassword/Slave_can_re-sync_with_master_after_password_change
> ```

ok, will inspect the unstable reason.

File: tests/gocase/unit/introspection/replication_test.go (hunk 1 above)
Comment (git-hulk) [reply tests/gocase/unit/introspection/replication_test.go]: @tisonkun I took a look the issue and it's Kvrocks's issue, so it's ok to keep this description.

File: tests/gocase/unit/introspection/replication_test.go (hunk 2 above)
Comment (git-hulk) [reply tests/gocase/unit/introspection/replication_test.go]: done

File: tests/gocase/unit/introspection/replication_test.go (hunk 5 above)
Comment (tisonkun) [reply tests/gocase/unit/introspection/replication_test.go]: @git-hulk I think you still hold the test under:

* `gocase/unit`

instead of:

* `gocase/integration`.

Comment (git-hulk) [reply tests/gocase/unit/introspection/replication_test.go]: @tisonkun it's in [tests/gocase/unit/replication/replication_test.go](https://github.com/apache/incubator-kvrocks/pull/997/files#diff-99ea2797442d2eaf01c9416f031da8c8089fa5725188e2b4f1a961d4572a9f7a) indeed, do you mean should move into the `integration` dir?

File: tests/gocase/unit/replication/replication_test.go [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+
+import (
+	"context"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/go-redis/redis/v9"
+
```
Comment (torwig) [inline tests/gocase/unit/replication/replication_test.go]: Redundant new line

File: tests/gocase/unit/replication/replication_test.go [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+	"github.com/stretchr/testify/require"
+
```
Comment (torwig) [inline tests/gocase/unit/replication/replication_test.go]: Redundant new line. @git-hulk If you sort imports, the `"github.com/apache/incubator-kvrocks/tests/gocase/util"` will be moved before `"github.com/go-redis/redis/v9"`

**Review by torwig — COMMENTED** [review]
LGTM, just imports should be sorted

File: tests/gocase/unit/introspection/replication_test.go (hunk 5 above)
Comment (tisonkun) [reply tests/gocase/unit/introspection/replication_test.go]: Yes. Replication tests originality lays under tcl/tests/integration.

File: tests/gocase/unit/replication/replication_test.go (hunk 6 above)
Comment (git-hulk) [reply tests/gocase/unit/replication/replication_test.go]: removed the redundant new line

File: tests/gocase/unit/introspection/replication_test.go (hunk 5 above)
Comment (git-hulk) [reply tests/gocase/unit/introspection/replication_test.go]: Oops, I misremembered the dir when migrating.

**Review by tisonkun — APPROVED** [review]
